### ISSUE TITLE
Can't Access AppRtc over the internet

### ISSUE THREAD

**Issue by charbelhass** [webrtc/apprtc#266]
Hello,

I was able to run AppRtc successfully on my public server after following the steps provided on github (https://github.com/webrtc/apprtc). 
(http://videochat.cloudapp.net:9090 , you can try it on firefox, it might not work on chrome because SSL is not implemented yet).

If the 2 devices are on the same network, a video call can be made without error. But if the 2 devices are not on the same network, they can't communicate with each other.
Please Help.

Thanks in advance.

**Comment by KaptenJansson** [comment]
Hi,

You need to setup a turn/stun server that provide connectivity checks for your clients to that they talk to each other on different networks.

We use [rfc5766-turn-server](https://github.com/coturn/rfc5766-turn-server) for this (it's only accessible from appr.tc and localhost:8080 hence you can only use the Google provided turn server when developing locally).

**Comment by charbelhass** [comment]
@KaptenJansson Thanks I managed to create A turn/stun server
but i have a question,
where should i put the parameter in Apprtc.
I mean the address port username and password of turn/stun server, where should I put them in Apprtc in order to work?

Thanks in advance

**Comment by KaptenJansson** [comment]
Since AppRTC gets those credentials from server in a JSON response from a CEOD server you would need to either have a web server doing the same or hard code those values in https://github.com/webrtc/apprtc/blob/master/src/app_engine/constants.py#L15 and then change https://github.com/webrtc/apprtc/blob/master/src/app_engine/apprtc.py#L256 to `turn_url = constants.TURN_URL_TEMPLATE`

**Comment by ubuntutest** [comment]
@KaptenJansson 
Hello, I really need help otherwise risk of becoming crazy. I can not figure out how to configure apprtc, I'm trying for so many days.
The app works correctly both locally and on a VPS.
I start it like this: dev_appserver.py ./out/app_engine/
The problem is the Turn server.
I followed countless guides without solving.
Turn / Stun
rfc5766-turn-server
Coturn ..
And more.
Keeps telling me "No turn server ..."
After several attempts I am trying to start the Turn server with this command:
`"sudo -v turnserver -a -o -n --no-dtls --no-tls -u test:test -r "someRealm"`
It should be as easy.
If i start this script
`function checkTURNServer(turnConfig, timeout){ 

  return new Promise(function(resolve, reject){

    setTimeout(function(){
        if(promiseResolved) return;
        resolve(false);
        promiseResolved = true;
    }, timeout || 5000);

    var promiseResolved = false
      , myPeerConnection = window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection   //compatibility for firefox and chrome
      , pc = new myPeerConnection({iceServers:[turnConfig]})
      , noop = function(){};
    pc.createDataChannel("");    //create a bogus data channel
    pc.createOffer(function(sdp){
      if(sdp.sdp.indexOf('typ relay') > -1){ // sometimes sdp contains the ice candidates...
        promiseResolved = true;
        resolve(true);
      }
      pc.setLocalDescription(sdp, noop, noop);
    }, noop);    // create offer and set local description
    pc.onicecandidate = function(ice){  //listen for candidate events
      if(promiseResolved || !ice || !ice.candidate || !ice.candidate.candidate || !(ice.candidate.candidate.indexOf('typ relay')>-1))  return;
      promiseResolved = true;
      resolve(true);
    };
  });   
}

checkTURNServer({
    url: 'turn:localhost',
    username: 'test',
    credential: 'test'
}).then(function(bool){
    console.log('is my TURN server active? ', bool? 'yes':'no');
}).catch(console.error.bind(console));`

He tells me that the turnserver is active.
Now where do I enter the Turn server settings? Really I can not understand, you have already answered this question in the previous comment, but I can not understand.
Change: TURN_BASE_URL = 'https://computeengineondemand.appspot.com' with
TURN_BASE_URL = 'https: // localhost'?
Other parameters such as CEOD_KEY leave it as it is?
Where do I enter the username and password?

Also in the line of code in 256 'apprtc.py there is nothing, I do not know what to change.

I ask you kindly to answer,
I'm sure many other people are in the same situation.
Thanks in advance

**Comment by KaptenJansson** [comment]
As I wrote, AppRTC does not connect to the TURN server in order to get a TURN server, it connects to an ICE server (STUN/TURN) provider (CEOD) which responds with a JSON format containing STUN and TURN servers and credentials. This is done in order to get active servers that are close to the person making the request, it actually fetches/starts a TURN server near you and provides the IP address and credentials for that server in the response. 

This has been recently replaced with a different service called network traversal which uses the ICE, do note that these are Google only services.

If you do not want this feature you can just add the turn server to directly to https://github.com/webrtc/apprtc/blob/master/src/web_app/js/call.js#L386.

e.g. 
```
var servers = [
  {
    credentials: "Your turn server password",
    urls: [turn:IP TO YOUR TURN SERVER:19305?transport=udp"],
    username: 'Your turn server username' 
  },
  {urls: ["stun:stun.l.google.com:19302"]}
];
```

**Comment by ubuntutest** [comment]
Thanks, I think I understand your explanation, but I have not yet figured out how to implement this feature, I have difficulty understanding how to use CEOD, but if there is an alternative, its ok.

I want to use the code you've shown me, but there is a small problem, I run the app from the /out/app_engine, and in this folder does not exist call.js.

If I try to run it from src/*, the app does not work.

Excuse me are not very experienced,
Thanks, I await response

**Comment by KaptenJansson** [comment]
I pointed to the source code, i.e. before it's compiled into a single file. You should find it in the src/ folder, change it as I suggested and the run grunt build again.

**Comment by ubuntutest** [comment]
You made me understand many things,

Use this command to start the server turn:
`sudo -v -a -o -n --no-DTLS --no-tls -u test: test -r "someRealm"`
If I run the script I linked before, he tells me that the Turnserver is active.
```

[...]
checkTURNServer({
    url: 'turn:localhost',
    username: 'test',
    credential: 'test'
}).then(function(bool){
    console.log('is my TURN server active? ', bool? 'yes':'no');
}).catch(console.error.bind(console));
Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}
VM115:36
 is my TURN server active?  yes

```
So I edit call.js with these parameters:

```
var servers = [
   {
     credentials: "test",
     urls: ["turn: localhost: 19305? transport = udp"],
     username: 'test'
   },
   {Urls: ["stun: stun.l.google.com: 19302"]}
];
```

Run the grunt build from the root dir and compile without error.
Starting from /out/app_engine
I go to localhost:8080,
"No Turn Server ..."
What am I doing wrong?

**Comment by ubuntutest** [comment]
Hey friend, do not abandon me now, I think we are on track to solve :)

**Comment by KaptenJansson** [comment]
Is the turn server listening on port 19305?

**Comment by KaptenJansson** [comment]
You can test your turn server using https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/

**Comment by ubuntutest** [comment]
Settings:
```
Turn URL > turn:localhost
Username > test
Credential > test
```

`turn:localhost[test:test]`

```
Time	Component	Type	Foundation	Protocol	Address	Port	Priority
0.003	1	host	2224966167	udp	192.168.1.68	52175	126 | 30 | 255
0.004	2	host	2224966167	udp	192.168.1.68	42131	126 | 30 | 254
0.006	1	srflx	842163049	udp	127.0.0.1	52175	100 | 30 | 255
0.006	1	srflx	4231669940	udp	::1	36895	100 | 60 | 255
0.006	2	srflx	842163049	udp	127.0.0.1	42131	100 | 30 | 254
0.006	2	srflx	4231669940	udp	::1	45142	100 | 60 | 254
0.070	1	relay	213309110	udp	127.0.0.1	62293	2 | 30 | 255
0.071	2	relay	2180157977	udp	::1	58925	2 | 60 | 254
0.071	1	relay	2180157977	udp	::1	55691	2 | 60 | 255
0.071	2	relay	213309110	udp	127.0.0.1	53234	2 | 30 | 254
0.153	Done

```

Thats the result, seems to work?
Sorry if i'm annoying you

**Comment by KaptenJansson** [comment]
So you're not specifying a port? Then change that in call.js then.

**Comment by ubuntutest** [comment]
No, i'm not specifying a port,
i've tested it too

```
var servers = [
   {
     credentials: "test",
     urls: ["turn:localhost"],
     username: 'test'
   },
   {Urls: ["stun:stun.l.google.com:19302"]}
];
```
But it still say "No turn server..."
Maybe i have to change something in constants.py? or .. i really don't know.

**Comment by KaptenJansson** [comment]
Have you checked the javascript logs in the browser?

**Comment by ubuntutest** [comment]
The console.log is very long, I think these are the important parts,
```
POST https://networktraversal.googleapis.com/v1alpha/iceconfig?key=none 400 
()(anonymous function) @ 
apprtc.debug.js:3955sendUrlRequest @ 
apprtc.debug.js:3936sendAsyncUrlRequest @
 apprtc.debug.js:3933(anonymous function) @
 apprtc.debug.js:3963requestIceServers @ 
apprtc.debug.js:3962Call.maybeGetIceServers_ @ apprtc.debug.js:2445Call.requestMediaAndIceServers_ @ 
apprtc.debug.js:2254Call @
 apprtc.debug.js:2250AppController.createCall_ @ 
apprtc.debug.js:1895(anonymous function) @
 apprtc.debug.js:1928RoomSelection.loadRoom_ @ apprtc.debug.js:3330RoomSelection.onJoinButton_ @
 apprtc.debug.js:3319
apprtc.debug.js:4055 6.048: ICE server request error: Status=400, response={
  "error": {
    "code": 400,
    "message": "API key not valid. Please pass a valid API key.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.Help",
        "links": [
          {
            "description": "Google developers console",
            "url": "https://console.developers.google.com"
          }
        ]
      }
    ]
  }
}

callstats.min.js:3 addNewFabric: SDK not initialized.


collector.callstats.io/o/authorize:1 POST https://collector.callstats.io/o/authorize 403 (Forbidden)
apprtc.debug.js:4055 9.190: Init status: httpError msg: HTTP 403, Forbidden. {"authorize":{"status":"Error","reason":"Wrong application identifier given."}}
```


FULL log here:
http://pastebin.com/YD4x1RTy

**Comment by KaptenJansson** [comment]
You have not set up your own signalling server. You cannot use the on that is hosted by Google.

I suggest you read the [Deployment instructions](https://github.com/webrtc/apprtc#deployment) again.

**Comment by ubuntutest** [comment]
You right, i'm sorry for this.
First time that worked, finally no more problems with turn server, but..

Now does not work anymore, I can not understand why, it's amazing.
What about the variable "roomSrv" I have to change it?
I entered localhost

`var = roomSrv flag.String ("room-server", "http://localhost", "The Origin of the room" server)`

This is the result of the collider:

```
$ $ GOPATH / bin / collidermain -port = 8088 -TLS = false
07.28.2016 18:56:38 Starting collider: tls = false, port = 8088, room-server = http: // localhost
```

Another strange thing:
When i run go test collider
sometimes it say "ok" sometimes "fail"

**Comment by ubuntutest** [comment]
My consolelog seems to be the same.

http://pastebin.com/FjRzbpUD

Of course its active;

```
$ $GOPATH/bin/collidermain -port=8089 -tls=false
2016/07/28 23:06:12 Starting collider: tls = false, port = 8089, room-server=http://localhost
```

**Comment by KaptenJansson** [comment]
You have not changed the Collider host address, it's still pointing to "apprtc-ws.webrtc.org". change it in constants.py. Also since you have disabled TLS on collider (which is bad idea for production but OK for testing) you need to change the protocol from wss:// to ws://.

**Comment by ubuntutest** [comment]
Changed apprtc-ws.webrtc.org to localhost

but from the console:
`WebSocket connection to 'wss://localhost/ws' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED`
How do i remove "wss" "/ws"

Thats my constant.py
http://pastebin.com/7ufGS09g

**Comment by KaptenJansson** [comment]
Add the ?wstls=false parameter when conencting to apprtc, e.g. http://localhost:8080/?wstls=false

**Comment by ubuntutest** [comment]
I've changed in main.go
`var tls = flag.Bool("tls", false, "whether TLS is used")`
Ad modified that:
```
  if wss_tls and wss_tls == 'false':
    wss_url = 'ws://' + wss_host_port_pair 
    wss_post_url = 'http://' + wss_host_port_pair
  else:
    wss_url = 'ws://' + wss_host_port_pair 
    wss_post_url = 'http://' + wss_host_port_pair
  return (wss_url, wss_post_url)
```
 Using 
`http://localhost:8083/?wstls=false`
Still don't work..

`WebSocket connection to 'ws://localhost/' failed: Error during WebSocket handshake: Unexpected response code: 200`

**Comment by KaptenJansson** [comment]
You have not added the port to the collider address which is 8089 according to your [comment](https://github.com/webrtc/apprtc/issues/266#issuecomment-236025841)

Sorry, I'm going to end it here, I cannot help you forever, these are fundamental things (you always need to specify a port etc) that one should know when trying to deploy a web application.

I will in the future try to improve on the documentation as it's not perfect but not sure when.

**Comment by ubuntutest** [comment]
Before you ask here in the forum, I always make several attempts.
I'm already using port 8089,
```
$ $ GOPATH / bin / collidermain -port = 8089 -TLS = false
29.07.2016 12:40:09 Starting collider: tls = false, port = 8089, room-server = http: // localhost: 8089

```
I also tried to change
`WSS_INSTANCE_HOST_KEY: 'localhost',`
in
`WSS_INSTANCE_HOST_KEY: 'localhost: 8089'`

But the problem is not solved
`WebSocket connection to 'ws: // localhost: 8089 /' failed: Error during WebSocket handshake: Unexpected response code: 500`

Thanks for your help, you were really kind, I understood many things, although I have not solved.
Unfortunately the documentation is useless, says nothing.

**Comment by pieronov** [comment]
Hello everyone, I am writing in this post because this helped me a lot, I hope its ok well even if it is closed.

I'm trying to use this app on a VPS, so I'm not using localhost, and I do not even https, this is the address:
`http://51.254.115.249:8083?wstls=false`

The collider is working properly, the log shows me the room opened, closed, connections, etc

To start the turnserver I also use the simple command
`sudo -v -a -o -n --no-DTLS --no-tls -u hello:hello -r someRealm`
Open ports are the 3478 and 3479, so I have included this variable in call.js file

`var servers = [{credentials: "hello", urls: [ "turn: 51254115249: 3478? transport = udp", "turn: 51254115249: 3478? transport = tcp"], username "hello"}, {urls: [ "stun: stun.l.google.com: 19302"]}];`

I also tried to insert it directly into apprtc.debug.js and utils.js file
But the TURN server is still down,
When I try to connect with your smartphone it does not work.

"No TURN server; unlikely That average will traverse networks"

Any tips for me? thank you
